src/common/s3_adapter.h

```diff
@@ -332,6 +333,19 @@ class S3Adapter {
         std::condition_variable cond_;
     };
 
+    class AsyncCallbackThreadPool : public TaskThreadPool<> {
```

Comment:
it seems this derived class is unnecessary, we can add an extra parameter to control whether run all tasks when task pool stopped.

```
    virtual void ThreadFunc() {
        while (running_.load(std::memory_order_acquire)
|| (drain && !queue.empty())
) {
            Task task(Take());
            /* ThreadPool 退出的时候，queue 为空，那么会返回无效的 task */
            if (task) {
                task();
            }
        }
    }
```